File: src/main/java/com/baidu/algorithm/dfs/_547_Friend_Circles.py
class Solution(object):
    def findCircleNum(self, M):
        """
        :type M: List[List[int]]
        :rtype: int
        """
        if not M or not M[0]:
            return 0

        n = len(M)
        uf = UnionFind(n)
        for i in xrange(n):
            for j in xrange(i + 1, n):
                if M[i][j]:
                    uf.union(i, j)

        return uf.num_of_class()


class UnionFind(object):
    def __init__(self, n):
        self.n = n
        self.father = range(n)
        self.rank = [1] * n

    def union(self, x, y):
        fx = self.find(x)
        fy = self.find(y)
        if fx != fy:
            if self.rank[fx] < self.rank[fy]:
                self.father[fx] = fy
            elif self.rank[fx] > self.rank[fy]:
                self.father[fy] = fx
            else:
                self.father[fy] = fx
                self.rank[fx] += 1

    def find(self, x):
        if x != self.father[x]:
            self.father[x] = self.find(self.father[x])

        return self.father[x]

    def num_of_class(self):
        father_set = set()
        for x in xrange(self.n):
            father_set.add(self.find(x))

        return len(father_set)


if __name__ == '__main__':
    s = Solution()
    s.findCircleNum([[1, 0, 0, 1], [0, 1, 1, 0], [0, 1, 1, 1], [1, 0, 1, 1]])

File: src/main/java/com/baidu/algorithm/dfs/_547_Friend_Circles.java
/*
 * Copyright (C) 2018 Baidu, Inc. All Rights Reserved.
 */
package com.baidu.algorithm.dfs;

/**
 * _547_Friend_Circles
 *
 * @author xuhaoran01
 */
public class _547_Friend_Circles {

    private void dfs(int[][] M, boolean[] found, int i) {
        if (found[i]) {
            return;
        }

        found[i] = true;
        for (int j = 0; j < M[i].length; j++) {
            if (M[i][j] == 1) {
                dfs(M, found, j);
            }
        }
    }

    public int findCircleNum1(int[][] M) {
        int n = M.length, res = 0;
        boolean[] found = new boolean[n];
        for (int i = 0; i < n; i++) {
            if (!found[i]) {
                res++;
                dfs(M, found, i);
            }
        }

        return res;
    }

    public int findCircleNum(int[][] M) {
        int n = M.length;
        UnionFind uf = new UnionFind(n);
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (M[i][j] == 1) {
                    uf.union(i, j);
                }
            }
        }

        return uf.size;
    }
}

class UnionFind {
    public int size;
    public int[] father;

    public UnionFind(int n) {
        size = n;

        father = new int[n];
        for (int i = 0; i < n; i++) {
            father[i] = i;
        }
    }

    public int find(int x) {
        if (father[x] != x) {
            father[x] = find(father[x]);
        }

        return father[x];
    }

    public void union(int x, int y) {
        int fx = find(x),
                fy = find(y);
        if (fx != fy) {
            father[fx] = fy;
            size--;
        }
    }
}
